src/storage/v2/property_store.cpp

```diff
@@ -1398,48 +1408,208 @@ uint32_t ToPowerOf8(uint32_t size) {
+}
+
+std::optional<utils::DecompressedBuffer> DecompressBuffer(DecodedBufferConst const &buffer_info) {
+  if (buffer_info.storage_mode != StorageMode::COMPRESSED) return std::nullopt;
+
+  // Memory (hex):
+  // 00 00 00 00 00 00 00
+  // |------------|         -> original size
+  //                ||      -> size modifier to get back to non multiple of 8
+  //                   |--- -> compressed data
+  // 0  1  2  3  4  5  6  (positions)
+
+  uint32_t original_size = 0;
+  auto const *data = buffer_info.view.data();
+  memcpy(&original_size, data, sizeof(uint32_t));
+
+  // we have to restore the original size of the compressed buffer + the size of the original buffer
+  auto modifier = data[sizeof(uint32_t)];
+  auto buffer_size = buffer_info.view.size_bytes();
+  auto compressed_size = (modifier != 0) ? (buffer_size - 8 + modifier) : buffer_size;
+
+  auto data_offset = sizeof(uint32_t) + 1;
+  auto compressed_buffer = std::span(data + data_offset, compressed_size - data_offset);
+  auto const *compressor = utils::Compressor::GetInstance();
+  auto decompressed_buffer = compressor->Decompress(compressed_buffer, original_size);
+
+  if (!decompressed_buffer) [[unlikely]] {
+    throw PropertyValueException("Failed to decompress buffer");
+  }
+
+  return decompressed_buffer;
+}
+
+void CompressBuffer(uint8_t (&buffer)[12], DecodedBuffer const &buffer_info) {
+  if (buffer_info.storage_mode != StorageMode::BUFFER) {
+    return;
+  }
+  auto uncompressed_size = buffer_info.view.size_bytes();
+
+  auto const *compressor = utils::Compressor::GetInstance();
+  auto compressed_buffer = compressor->Compress(buffer_info.view);
+  if (!compressed_buffer) {
+    throw PropertyValueException("Failed to compress buffer");
+  }
+
+  auto compressed_view = compressed_buffer->view();
+  auto metadata_size = sizeof(uint32_t) + 1;
+  auto size_needed = compressed_view.size_bytes() + metadata_size;
+  auto compressed_size_to_multiple_of_8 = ToMultipleOf8(size_needed);
+  if (compressed_size_to_multiple_of_8 >= uncompressed_size) {
+    // Compressed buffer + metadata are larger than the original buffer, so we don't perform the compression.
+    return;
+  }
+
+  auto compressed_data = std::make_unique_for_overwrite<uint8_t[]>(compressed_size_to_multiple_of_8);
+
+  // We have compressed data + new buffer to put it into, no need for old uncompressed buffer
+  FreeMemory(buffer_info);
+
+  // first 4 bytes are the size of the original buffer
+  auto orig_size = compressed_buffer->original_size();
+  memcpy(compressed_data.get(), &orig_size, sizeof(uint32_t));
+
+  // next byte is the mod before multiple of 8
+  const uint8_t mod = size_needed % 8;
+  compressed_data[sizeof(uint32_t)] = mod;
+
+  // the rest of the buffer is the compressed data
+  memcpy(compressed_data.get() + metadata_size, compressed_view.data(), compressed_view.size_bytes());
+
+  SetSizeData(buffer, compressed_size_to_multiple_of_8 + kUseCompressedBuffer, compressed_data.release());
```

Comment:
I think using `kUseCompressedBuffer` is correct, since we are using `metadata_size` in the calculation of `compressed_size_to_multiple_of_8`. We have to use `kUseCompressedBuffer` when updating the size since we are storing metadata that indicates that data is compressed in the size field with `kUseCompressedBuffer`.
@Ignition can you confirm if this is right?

Reply:
Yeah `compressed_size_to_multiple_of_8` is the actual allocated size. The `kUseCompressedBuffer` is just the lower 3 bits being set. Maybe would have been more obvious as `compressed_size_to_multiple_of_8 | kUseCompressedBuffer` to show we are bitwise inserting some information.